lib/FIR/flattened.h

```diff
@@ -0,0 +1,264 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_FIR_FLATTENED_H_
+#define FORTRAN_FIR_FLATTENED_H_
+
+#include "common.h"
+#include "mixin.h"
+#include "../parser/parse-tree.h"
+#include <list>
+#include <map>
+#include <set>
+#include <vector>
+
+namespace Fortran::FIR {
+
+struct AnalysisData;
+
+namespace flat {
+
+// This is a flattened, linearized representation of the parse tree. It captures
+// the executable specification of the input program. The flattened IR can be
+// used to construct the Fortran IR.
+
+struct Op;
+struct LabelOp;
+struct GotoOp;
+struct ReturnOp;
+struct ConditionalGotoOp;
+struct SwitchIOOp;
+struct SwitchOp;
+struct ActionOp;
+struct BeginOp;
+struct EndOp;
+struct IndirectGotoOp;
+struct DoIncrementOp;
+struct DoCompareOp;
+
+using LabelRef = unsigned;
+constexpr LabelRef unspecifiedLabel{~0u};
+
+using Location = parser::CharBlock;
+struct LabelBuilder;
+
+// target for a control-flow edge
+struct LabelOp {
+  explicit LabelOp(LabelBuilder &builder);
+  LabelOp(const LabelOp &that);
+  LabelOp &operator=(const LabelOp &that);
+  void setReferenced() const;
+  bool isReferenced() const;
+  LabelRef get() const { return label_; }
+  operator LabelRef() const { return get(); }
+  void dump() const;
```

Comment:
Follow our naming conventions, these functions should be `SetReferenced()`, `IsReferenced()`, `Get()` (or `label()`), and `Dump()`.

Reply:
Quoting from the project C++Style.md:

**In Brief**
> - Where **LLVM's** C++ style guide is clear on usage, follow it.

**Naming***
> 1. C++ names that correspond to well-known interfaces from the STL and **LLVM** can and should look like their models when the reader can safely assume that they mean the same thing -- e.g., clear() and size() member functions in a class that implements an STL-ish container.

Please fix the style guide so it isn't ambiguous.

Reply:
I will make the style guide more clear.  Consistency with established precedents (now that we have them) is more important.

Reply:
I'm going to decline to change the names of all the member functions on the grounds that they conform with the LLVM coding standard from which they came, do not pose any significant challenge to understanding the code, and latitude is granted to the programmer in the F18 style guide.